GTA02-psuedo support.

---
commit 9b2c7de6c48b66346a719c476b218a1b9cbe47a0
tree b51c086e9869298a4073895d635da80c2df7ea6c
parent fbfbc5130075e233f2fd2613e2897c3c4ed4bc7f
author Andrzej Zaborowski <balrog@zabor.org> Tue, 27 May 2008 14:37:06 +0200
committer Andrzej Zaborowski <balrog@zabor.org> Tue, 27 May 2008 14:37:06 +0200

 arch/arm/boot/compressed/Makefile      |    4 
 arch/arm/boot/compressed/head-qemu.S   |   32 +
 arch/arm/mach-s3c2410/Kconfig          |    8 
 arch/arm/mach-s3c2410/Makefile         |    1 
 arch/arm/mach-s3c2410/mach-gta01.c     |   14 -
 arch/arm/mach-s3c2410/mach-gta02f.c    |  811 ++++++++++++++++++++++++++++++++
 arch/arm/mach-s3c2440/mach-gta02.c     |   24 -
 arch/arm/tools/mach-types              |    1 
 defconfig-2.6.22.5                     |    1 
 defconfig-2.6.24                       |    1 
 defconfig-gta02                        |    1 
 drivers/sdio/hcd/s3c24xx/s3c24xx_hcd.c |   44 +-
 include/asm-arm/plat-s3c24xx/pm.h      |    2 
 13 files changed, 903 insertions(+), 41 deletions(-)

diff --git a/arch/arm/boot/compressed/Makefile b/arch/arm/boot/compressed/Makefile
index 5fde99f..ac32d45 100644
--- a/arch/arm/boot/compressed/Makefile
+++ b/arch/arm/boot/compressed/Makefile
@@ -48,6 +48,10 @@ ifeq ($(CONFIG_ARCH_AT91RM9200),y)
 OBJS		+= head-at91rm9200.o
 endif
 
+ifeq ($(CONFIG_MACH_NEO1973_GTA02F),y)
+OBJS		+= head-qemu.o
+endif
+
 ifeq ($(CONFIG_CPU_BIG_ENDIAN),y)
 ifeq ($(CONFIG_CPU_CP15),y)
 OBJS		+= big-endian.o
diff --git a/arch/arm/boot/compressed/head-qemu.S b/arch/arm/boot/compressed/head-qemu.S
new file mode 100644
index 0000000..abc1901
--- /dev/null
+++ b/arch/arm/boot/compressed/head-qemu.S
@@ -0,0 +1,32 @@
+/*
+ * arch/arm/boot/compressed/head-qemu.S
+ *
+ * Copyright (C) 2007 OpenMoko Inc.
+ *
+ * We want to check here whether we are running in the emulator, on
+ * an emulated machine that is equipped with more stuff than the one
+ * claimed by the machine ID from the bootloader.  In such case we
+ * override the machine ID, so that the we can still use standard
+ * bootloaders.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include <linux/linkage.h>
+
+	.section        ".start", "ax"
+__qemu_start:
+
+	/*
+	 * Check that the virtual register has the value it should have
+	 * and if so, write the number of register in which we want to
+	 * receive the machine ID, to our virtual register.  That's it.
+	 */
+	mov	r0, #0x28000000		@ Base address of machine ID register
+	mov	r1, #7			@ Our machine ID is in r7
+	orr	r1, r1, #('Q' << 4)	@ OR it with magic value 1
+	ldr	r2, [r0]		@ Load the register value
+	cmp	r2, #'Q'		@ Compare against magic value 2
+	streq	r1, [r0]		@ If ok, get the machine ID
diff --git a/arch/arm/mach-s3c2410/Kconfig b/arch/arm/mach-s3c2410/Kconfig
index e42195f..843ecbd 100644
--- a/arch/arm/mach-s3c2410/Kconfig
+++ b/arch/arm/mach-s3c2410/Kconfig
@@ -125,5 +125,13 @@ config MACH_NEO1973_GTA01
 	help
 	   Say Y here if you are using the FIC Neo1973 GSM Phone
 
+config MACH_NEO1973_GTA02F
+	bool "QEMU Neo1973 GSM Phone (Paravirtual GTA02 Hardware)"
+	select CPU_S3C2410
+	select SENSORS_PCF50606
+	help
+	   Say Y here if you are using the QEMU Neo1973 GTA02-like
+	   GSM phone by running qemu with "-M gta02fake" commandline.
+
 endmenu
 
diff --git a/arch/arm/mach-s3c2410/Makefile b/arch/arm/mach-s3c2410/Makefile
index bb577f4..5208bb0 100644
--- a/arch/arm/mach-s3c2410/Makefile
+++ b/arch/arm/mach-s3c2410/Makefile
@@ -31,3 +31,4 @@ obj-$(CONFIG_BAST_PC104_IRQ)	+= bast-irq.o
 obj-$(CONFIG_MACH_VR1000)	+= mach-vr1000.o usb-simtec.o
 obj-$(CONFIG_MACH_QT2410)	+= mach-qt2410.o
 obj-$(CONFIG_MACH_NEO1973_GTA01)+= mach-gta01.o
+obj-$(CONFIG_MACH_NEO1973_GTA02F)+= mach-gta02f.o
diff --git a/arch/arm/mach-s3c2410/mach-gta01.c b/arch/arm/mach-s3c2410/mach-gta01.c
index b0cb544..99c36e6 100644
--- a/arch/arm/mach-s3c2410/mach-gta01.c
+++ b/arch/arm/mach-s3c2410/mach-gta01.c
@@ -272,7 +272,7 @@ static struct resource gta01_pmu_resources[] = {
 	},
 };
 
-struct platform_device gta01_pmu_dev = {
+static struct platform_device gta01_pmu_dev = {
 	.name 		= "pcf50606",
 	.num_resources	= ARRAY_SIZE(gta01_pmu_resources),
 	.resource	= gta01_pmu_resources,
@@ -505,13 +505,13 @@ static struct s3c2410_ts_mach_info gta01_ts_cfg = {
 
 /* SPI */
 
-void gta01_jbt6k74_reset(int devidx, int level)
+static void gta01_jbt6k74_reset(int devidx, int level)
 {
 	/* empty place holder; gta01 does not yet use this */
 	printk(KERN_DEBUG "gta01_jbt6k74_reset\n");
 }
 
-const struct jbt6k74_platform_data gta01_jbt6k74_pdata = {
+static const struct jbt6k74_platform_data gta01_jbt6k74_pdata = {
 	.reset		= gta01_jbt6k74_reset,
 };
 
@@ -568,7 +568,7 @@ static struct resource s3c_spi_lcm_resource[] = {
 	},
 };
 
-struct platform_device s3c_device_spi_lcm = {
+static struct platform_device s3c_device_spi_lcm = {
 	.name		  = "spi_s3c24xx_gpio",
 	.id		  = 1,
 	.num_resources	  = ARRAY_SIZE(s3c_spi_lcm_resource),
@@ -591,7 +591,7 @@ static struct resource gta01_bl_resources[] = {
 	},
 };
 
-struct platform_device gta01_bl_dev = {
+static struct platform_device gta01_bl_dev = {
 	.name 		= "gta01-bl",
 	.num_resources	= ARRAY_SIZE(gta01_bl_resources),
 	.resource	= gta01_bl_resources,
@@ -607,7 +607,7 @@ static struct resource gta01_led_resources[] = {
 	},
 };
 
-struct platform_device gta01_led_dev = {
+static struct platform_device gta01_led_dev = {
 	.name		= "neo1973-vibrator",
 	.num_resources	= ARRAY_SIZE(gta01_led_resources),
 	.resource	= gta01_led_resources,
@@ -628,7 +628,7 @@ static struct resource gta01_button_resources[] = {
 	},
 };
 
-struct platform_device gta01_button_dev = {
+static struct platform_device gta01_button_dev = {
 	.name		= "neo1973-button",
 	.num_resources	= ARRAY_SIZE(gta01_button_resources),
 	.resource	= gta01_button_resources,
diff --git a/arch/arm/mach-s3c2410/mach-gta02f.c b/arch/arm/mach-s3c2410/mach-gta02f.c
new file mode 100644
index 0000000..a51f8e4
--- /dev/null
+++ b/arch/arm/mach-s3c2410/mach-gta02f.c
@@ -0,0 +1,811 @@
+/*
+ * linux/arch/arm/mach-s3c2410/mach-gta02f.c
+ *
+ * S3C2410 Machine Support for the paravirtual GTA02 in qemu-neo1973.
+ * All its specs are that of GTA01 except the following peripherals:
+ *  o the GPS chip on UART1
+ *  o the WLAN chip on SDIO
+ *  o the accelerometers
+ *  o no SD slot
+ *  o third UART
+ *  o TODO: Flash size
+ *  o TODO: "-version" and "-resume" devices
+ *
+ * Copyright (C) 2007 OpenMoko, Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+ * MA 02111-1307 USA
+ *
+ */
+
+#include <linux/kernel.h>
+#include <linux/types.h>
+#include <linux/interrupt.h>
+#include <linux/list.h>
+#include <linux/timer.h>
+#include <linux/init.h>
+#include <linux/workqueue.h>
+#include <linux/platform_device.h>
+#include <linux/serial_core.h>
+#include <asm/arch/ts.h>
+#include <linux/spi/spi.h>
+#include <linux/spi/spi_bitbang.h>
+#include <linux/mmc/mmc.h>
+#include <linux/mmc/host.h>
+
+#include <linux/mtd/mtd.h>
+#include <linux/mtd/nand.h>
+#include <linux/mtd/nand_ecc.h>
+#include <linux/mtd/partitions.h>
+
+#include <linux/mmc/host.h>
+
+#include <linux/pcf50606.h>
+
+#include <asm/mach/arch.h>
+#include <asm/mach/map.h>
+#include <asm/mach/irq.h>
+
+#include <asm/hardware.h>
+#include <asm/io.h>
+#include <asm/irq.h>
+#include <asm/mach-types.h>
+
+#include <asm/arch/regs-gpio.h>
+#include <asm/arch/fb.h>
+#include <asm/arch/mci.h>
+#include <asm/arch/spi.h>
+#include <asm/arch/spi-gpio.h>
+#include <asm/arch/usb-control.h>
+
+#include <asm/arch/gta01.h>
+#include <asm/arch/gta02.h>
+
+#include <asm/plat-s3c/regs-serial.h>
+#include <asm/plat-s3c/nand.h>
+#include <asm/plat-s3c24xx/devs.h>
+#include <asm/plat-s3c24xx/cpu.h>
+#include <asm/plat-s3c24xx/pm.h>
+#include <asm/plat-s3c24xx/udc.h>
+#include <asm/plat-s3c24xx/neo1973.h>
+#include <linux/jbt6k74.h>
+
+#include <linux/neospy.h>
+extern struct neospy_logdata neospy;
+
+static struct map_desc gta01_iodesc[] __initdata = {
+	{
+		.virtual	= 0xe0000000,
+		.pfn		= __phys_to_pfn(S3C2410_CS3+0x01000000),
+		.length		= SZ_1M,
+		.type		= MT_DEVICE
+	},
+};
+
+#define UCON S3C2410_UCON_DEFAULT
+#define ULCON S3C2410_LCON_CS8 | S3C2410_LCON_PNONE | S3C2410_LCON_STOPB
+#define UFCON S3C2410_UFCON_RXTRIG8 | S3C2410_UFCON_FIFOMODE
+/* UFCON for the gta01 sets the FIFO trigger level at 4, not 8 */
+#define UFCON_GTA01_PORT0 S3C2410_UFCON_FIFOMODE
+
+static struct s3c2410_uartcfg gta01_uartcfgs[] = {
+	[0] = {
+		.hwport	     = 0,
+		.flags	     = 0,
+		.ucon	     = UCON,
+		.ulcon	     = ULCON,
+		.ufcon	     = UFCON_GTA01_PORT0,
+	},
+	[1] = {
+		.hwport	     = 1,
+		.flags	     = 0,
+		.ucon	     = UCON,
+		.ulcon	     = ULCON,
+		.ufcon	     = UFCON,
+	},
+	[2] = {
+		.hwport	     = 2,
+		.flags	     = 0,
+		.ucon	     = UCON,
+		.ulcon	     = ULCON,
+		.ufcon	     = UFCON,
+	},
+};
+
+/* PMU driver info */
+
+static int pmu_callback(struct device *dev, unsigned int feature,
+			enum pmu_event event)
+{
+	switch (feature) {
+	case PCF50606_FEAT_ACD:
+		switch (event) {
+		case PMU_EVT_INSERT:
+			pcf50606_charge_fast(pcf50606_global, 1);
+			break;
+		case PMU_EVT_REMOVE:
+			pcf50606_charge_fast(pcf50606_global, 0);
+			break;
+		default:
+			break;
+		}
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static struct pcf50606_platform_data gta01_pcf_pdata = {
+	.used_features	= PCF50606_FEAT_EXTON |
+			  PCF50606_FEAT_MBC |
+			  PCF50606_FEAT_BBC |
+			  PCF50606_FEAT_RTC |
+			  PCF50606_FEAT_WDT |
+			  PCF50606_FEAT_CHGCUR |
+			  PCF50606_FEAT_BATVOLT |
+			  PCF50606_FEAT_BATTEMP,
+	.onkey_seconds_required = 3,
+	.cb		= &pmu_callback,
+	.r_fix_batt	= 10000,
+	.r_fix_batt_par	= 10000,
+	.r_sense_milli	= 220,
+	.rails	= {
+		[PCF50606_REGULATOR_D1REG] = {
+			.name		= "bt_3v15",
+			.voltage	= {
+				.init	= 3150,
+				.max	= 3150,
+			},
+		},
+		[PCF50606_REGULATOR_D2REG] = {
+			.name		= "gl_2v5",
+			.voltage	= {
+				.init	= 2500,
+				.max	= 2500,
+			},
+		},
+		[PCF50606_REGULATOR_D3REG] = {
+			.name		= "stby_1v8",
+			.flags		= PMU_VRAIL_F_SUSPEND_ON,
+			.voltage	= {
+				.init	= 1800,
+				.max	= 2100,
+			},
+		},
+		[PCF50606_REGULATOR_DCD] = {
+			.name		= "gl_1v5",
+			.voltage	= {
+				.init	= 1500,
+				.max	= 1500,
+			},
+		},
+		[PCF50606_REGULATOR_DCDE] = {
+			.name		= "io_3v3",
+			.flags		= PMU_VRAIL_F_SUSPEND_ON,
+			.voltage	= {
+				.init	= 3300,
+				.max	= 3330,
+			},
+		},
+		[PCF50606_REGULATOR_DCUD] = {
+			.name		= "core_1v8",
+			.flags		= PMU_VRAIL_F_SUSPEND_ON,
+			.voltage	= {
+				.init	= 2100,
+				.max	= 2100,
+			},
+		},
+		[PCF50606_REGULATOR_IOREG] = {
+			.name		= "codec_3v3",
+			.voltage	= {
+				.init	= 3300,
+				.max	= 3300,
+			},
+		},
+		[PCF50606_REGULATOR_LPREG] = {
+			.name		= "lcm_3v3",
+			.voltage	= {
+				.init	= 3300,
+				.max	= 3300,
+			},
+		}
+	},
+};
+
+static void cfg_pmu_vrail(struct pmu_voltage_rail *vrail, char *name,
+			  unsigned int flags, unsigned int init,
+			  unsigned int max)
+{
+	vrail->name = name;
+	vrail->flags = flags;
+	vrail->voltage.init = init;
+	vrail->voltage.max = max;
+}
+
+static void mangle_pmu_pdata_by_system_rev(void)
+{
+	switch (system_rev) {
+	case GTA01Bv4_SYSTEM_REV:
+		gta01_pcf_pdata.used_features |= PCF50606_FEAT_ACD;
+		break;
+	case GTA01Bv3_SYSTEM_REV:
+	case GTA01Bv2_SYSTEM_REV:
+		gta01_pcf_pdata.rails[PCF50606_REGULATOR_D3REG]
+					.name = "user1";
+		gta01_pcf_pdata.rails[PCF50606_REGULATOR_D3REG]
+					.flags &= ~PMU_VRAIL_F_SUSPEND_ON;
+		gta01_pcf_pdata.rails[PCF50606_REGULATOR_D3REG]
+					.flags = PMU_VRAIL_F_UNUSED;
+		break;
+	case GTA01v4_SYSTEM_REV:
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_DCUD],
+			      "core_1v8", PMU_VRAIL_F_SUSPEND_ON, 1800, 1800);
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_D1REG],
+			      "vrf_3v", 0, 3000, 3000);
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_D3REG],
+			      "vtcxo_2v8", 0, 2800, 2800);
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_DCD],
+			      "gl_3v5", 0, 3500, 3500);
+		break;
+	case GTA01v3_SYSTEM_REV:
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_D1REG],
+			      "vrf_3v", 0, 3000, 3000);
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_D2REG],
+			      "sd_3v3", 0, 3300, 3300);
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_D3REG],
+			      "codec_3v3", 0, 3300, 3300);
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_DCD],
+			      "gpsio_3v3", 0, 3300, 3300);
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_DCUD],
+			      "core_1v8", PMU_VRAIL_F_SUSPEND_ON, 1800, 1800);
+		cfg_pmu_vrail(&gta01_pcf_pdata.rails[PCF50606_REGULATOR_IOREG],
+			      "vtcxo_2v8", 0, 2800, 2800);
+		break;
+	}
+}
+
+static struct resource gta01_pmu_resources[] = {
+	[0] = {
+		.flags	= IORESOURCE_IRQ,
+		.start	= GTA01_IRQ_PCF50606,
+		.end	= GTA01_IRQ_PCF50606,
+	},
+};
+
+static struct platform_device gta01_pmu_dev = {
+	.name 		= "pcf50606",
+	.num_resources	= ARRAY_SIZE(gta01_pmu_resources),
+	.resource	= gta01_pmu_resources,
+	.dev		= {
+		.platform_data = &gta01_pcf_pdata,
+	},
+};
+
+/* LCD driver info */
+
+/* Configuration for 480x640 toppoly TD028TTEC1.
+ * Do not mark this as __initdata or it will break! */
+static struct s3c2410fb_display gta01_displays[] =  {
+	{
+		.type		= S3C2410_LCDCON1_TFT,
+		.width		= 43,
+		.height		= 58,
+		.xres		= 480,
+		.yres		= 640,
+		.bpp		= 16,
+
+		.pixclock	= 40000,	/* HCLK/4 */
+		.left_margin	= 104,
+		.right_margin	= 8,
+		.hsync_len	= 8,
+		.upper_margin	= 2,
+		.lower_margin	= 16,
+		.vsync_len	= 2,
+		.lcdcon5	= S3C2410_LCDCON5_FRM565 |
+				  S3C2410_LCDCON5_INVVCLK |
+				  S3C2410_LCDCON5_INVVLINE |
+				  S3C2410_LCDCON5_INVVFRAME |
+				  S3C2410_LCDCON5_PWREN |
+				  S3C2410_LCDCON5_HWSWP,
+	},
+	{
+		.type		= S3C2410_LCDCON1_TFT,
+		.width		= 43,
+		.height		= 58,
+		.xres		= 480,
+		.yres		= 640,
+		.bpp		= 32,
+
+		.pixclock	= 40000,	/* HCLK/4 */
+		.left_margin	= 104,
+		.right_margin	= 8,
+		.hsync_len	= 8,
+		.upper_margin	= 2,
+		.lower_margin	= 16,
+		.vsync_len	= 2,
+		.lcdcon5	= S3C2410_LCDCON5_FRM565 |
+				  S3C2410_LCDCON5_INVVCLK |
+				  S3C2410_LCDCON5_INVVLINE |
+				  S3C2410_LCDCON5_INVVFRAME |
+				  S3C2410_LCDCON5_PWREN |
+				  S3C2410_LCDCON5_HWSWP,
+	},
+	{
+		.type		= S3C2410_LCDCON1_TFT,
+		.width		= 43,
+		.height		= 58,
+		.xres		= 240,
+		.yres		= 320,
+		.bpp		= 16,
+
+		.pixclock	= 40000,	/* HCLK/4 */
+		.left_margin	= 104,
+		.right_margin	= 8,
+		.hsync_len	= 8,
+		.upper_margin	= 2,
+		.lower_margin	= 16,
+		.vsync_len	= 2,
+		.lcdcon5	= S3C2410_LCDCON5_FRM565 |
+				  S3C2410_LCDCON5_INVVCLK |
+				  S3C2410_LCDCON5_INVVLINE |
+				  S3C2410_LCDCON5_INVVFRAME |
+				  S3C2410_LCDCON5_PWREN |
+				  S3C2410_LCDCON5_HWSWP,
+	},
+};
+
+static struct s3c2410fb_mach_info gta01_lcd_cfg __initdata = {
+	.displays	= gta01_displays,
+	.num_displays	= ARRAY_SIZE(gta01_displays),
+	.default_display = 0,
+
+	.lpcsel		= ((0xCE6) & ~7) | 1<<4,
+};
+
+static struct platform_device *gta01_devices[] __initdata = {
+	&s3c_device_usb,
+	&s3c_device_lcd,
+	&s3c_device_wdt,
+	&s3c_device_i2c,
+	&s3c_device_iis,
+	// &s3c_device_sdi, /* FIXME: temporary disable to avoid s3cmci bind */
+	&s3c_device_usbgadget,
+	&s3c_device_nand,
+	&s3c_device_ts,
+};
+
+static struct s3c2410_nand_set gta01_nand_sets[] = {
+	[0] = {
+		.name		= "neo1973-nand",
+		.nr_chips	= 1,
+		.flags		= S3C2410_NAND_BBT,
+	},
+};
+
+static struct s3c2410_platform_nand gta01_nand_info = {
+	.tacls		= 20,
+	.twrph0		= 60,
+	.twrph1		= 20,
+	.nr_sets	= ARRAY_SIZE(gta01_nand_sets),
+	.sets		= gta01_nand_sets,
+};
+
+static void gta01_mmc_set_power(unsigned char power_mode, unsigned short vdd)
+{
+	int bit;
+	int mv = 1700; /* 1.7V for MMC_VDD_165_195 */
+
+	printk(KERN_DEBUG "mmc_set_power(power_mode=%u, vdd=%u)\n",
+	       power_mode, vdd);
+
+	switch (system_rev) {
+	case GTA01v3_SYSTEM_REV:
+		switch (power_mode) {
+		case MMC_POWER_OFF:
+			pcf50606_onoff_set(pcf50606_global,
+					   PCF50606_REGULATOR_D2REG, 0);
+			break;
+		case MMC_POWER_ON:
+			/* translate MMC_VDD_* VDD bit to mv */
+			for (bit = 8; bit != 24; bit++)
+				if (vdd == (1 << bit))
+					mv += 100 * (bit - 4);
+			pcf50606_voltage_set(pcf50606_global,
+					     PCF50606_REGULATOR_D2REG, mv);
+			pcf50606_onoff_set(pcf50606_global,
+					   PCF50606_REGULATOR_D2REG, 1);
+			break;
+		}
+		break;
+	case GTA01v4_SYSTEM_REV:
+	case GTA01Bv2_SYSTEM_REV:
+	case GTA01Bv3_SYSTEM_REV:
+	case GTA01Bv4_SYSTEM_REV:
+		switch (power_mode) {
+		case MMC_POWER_OFF:
+			neo1973_gpb_setpin(GTA01_GPIO_SDMMC_ON, 1);
+			break;
+		case MMC_POWER_ON:
+			neo1973_gpb_setpin(GTA01_GPIO_SDMMC_ON, 0);
+			break;
+		}
+		break;
+	}
+}
+
+static struct s3c24xx_mci_pdata gta01_mmc_cfg = {
+	.gpio_detect	= GTA01_GPIO_nSD_DETECT,
+	.set_power	= &gta01_mmc_set_power,
+	.ocr_avail	= MMC_VDD_165_195|MMC_VDD_20_21|
+			  MMC_VDD_21_22|MMC_VDD_22_23|MMC_VDD_23_24|
+			  MMC_VDD_24_25|MMC_VDD_25_26|MMC_VDD_26_27|
+			  MMC_VDD_27_28|MMC_VDD_28_29|MMC_VDD_29_30|
+			  MMC_VDD_30_31|MMC_VDD_31_32|MMC_VDD_32_33,
+};
+
+static struct resource gta02_sdio_resources[] = {
+	[0] = {
+		.flags	= IORESOURCE_IRQ,
+		.start	= IRQ_SDI,
+		.end	= IRQ_SDI,
+	},
+	[1] = {
+		.flags = IORESOURCE_MEM,
+		.start = S3C2410_PA_SDI,
+		.end   = S3C2410_PA_SDI + S3C24XX_SZ_SDI - 1,
+	},
+	[2] = {
+		.flags = IORESOURCE_DMA,
+		.start = 0, /* Channel 0 for SDI */
+		.end = 0,
+	},
+};
+
+static struct platform_device gta02_sdio_dev = {
+        .name           = "s3c24xx-sdio",
+        .id             = -1,
+        .dev            = {
+                                .coherent_dma_mask      = 0xffffffff,
+        },
+        .resource       = gta02_sdio_resources,
+        .num_resources  = ARRAY_SIZE(gta02_sdio_resources),
+};
+
+static void gta01_udc_command(enum s3c2410_udc_cmd_e cmd)
+{
+	printk(KERN_DEBUG "%s(%d)\n", __func__, cmd);
+
+	switch (cmd) {
+	case S3C2410_UDC_P_ENABLE:
+		neo1973_gpb_setpin(GTA01_GPIO_USB_PULLUP, 1);
+		break;
+	case S3C2410_UDC_P_DISABLE:
+		neo1973_gpb_setpin(GTA01_GPIO_USB_PULLUP, 0);
+		break;
+	default:
+		break;
+	}
+}
+
+/* use a work queue, since I2C API inherently schedules
+ * and we get called in hardirq context from UDC driver */
+
+struct vbus_draw {
+	struct work_struct work;
+	int ma;
+};
+static struct vbus_draw gta01_udc_vbus_drawer;
+
+static void __gta01_udc_vbus_draw(struct work_struct *work)
+{
+	/* this is a fix to work around boot-time ordering problems if the
+	 * s3c2410_udc is initialized before the pcf50606 driver has defined
+	 * pcf50606_global */
+	if (!pcf50606_global)
+		return;
+
+	if (gta01_udc_vbus_drawer.ma >= 500) {
+		/* enable fast charge */
+		printk(KERN_DEBUG "udc: enabling fast charge\n");
+		pcf50606_charge_fast(pcf50606_global, 1);
+	} else {
+		/* disable fast charge */
+		printk(KERN_DEBUG "udc: disabling fast charge\n");
+		pcf50606_charge_fast(pcf50606_global, 0);
+	}
+}
+
+static void gta01_udc_vbus_draw(unsigned int ma)
+{
+	gta01_udc_vbus_drawer.ma = ma;
+	schedule_work(&gta01_udc_vbus_drawer.work);
+}
+
+static struct s3c2410_udc_mach_info gta01_udc_cfg = {
+	.vbus_draw	= gta01_udc_vbus_draw,
+};
+
+static struct s3c2410_ts_mach_info gta01_ts_cfg = {
+	.delay = 10000,
+	.presc = 65,
+	.oversampling_shift = 5,
+};
+
+/* SPI */
+
+static void gta01_jbt6k74_reset(int devidx, int level)
+{
+	/* empty place holder; gta01 does not yet use this */
+	printk(KERN_DEBUG "gta01_jbt6k74_reset\n");
+}
+
+static const struct jbt6k74_platform_data gta01_jbt6k74_pdata = {
+	.reset		= gta01_jbt6k74_reset,
+};
+
+static struct spi_board_info gta01_spi_board_info[] = {
+	{
+		.modalias	= "jbt6k74",
+		.platform_data	= &gta01_jbt6k74_pdata,
+		/* controller_data */
+		/* irq */
+		.max_speed_hz	= 10 * 1000 * 1000,
+		.bus_num	= 1,
+		/* chip_select */
+	},
+};
+
+static void spi_gpio_cs(struct s3c2410_spigpio_info *spi, int csidx, int cs)
+{
+	switch (cs) {
+	case BITBANG_CS_ACTIVE:
+		s3c2410_gpio_setpin(S3C2410_GPG3, 0);
+		break;
+	case BITBANG_CS_INACTIVE:
+		s3c2410_gpio_setpin(S3C2410_GPG3, 1);
+		break;
+	}
+}
+
+static struct s3c2410_spigpio_info spi_gpio_cfg = {
+	.pin_clk	= S3C2410_GPG7,
+	.pin_mosi	= S3C2410_GPG6,
+	.pin_miso	= S3C2410_GPG5,
+	.board_size	= ARRAY_SIZE(gta01_spi_board_info),
+	.board_info	= gta01_spi_board_info,
+	.chip_select	= &spi_gpio_cs,
+	.num_chipselect = 2,   /*** Should be 1 or 2 for gta01? ***/
+};
+
+static struct resource s3c_spi_lcm_resource[] = {
+	[0] = {
+		.start = S3C2410_GPG3,
+		.end   = S3C2410_GPG3,
+	},
+	[1] = {
+		.start = S3C2410_GPG5,
+		.end   = S3C2410_GPG5,
+	},
+	[2] = {
+		.start = S3C2410_GPG6,
+		.end   = S3C2410_GPG6,
+	},
+	[3] = {
+		.start = S3C2410_GPG7,
+		.end   = S3C2410_GPG7,
+	},
+};
+
+static struct platform_device s3c_device_spi_lcm = {
+	.name		  = "spi_s3c24xx_gpio",
+	.id		  = 1,
+	.num_resources	  = ARRAY_SIZE(s3c_spi_lcm_resource),
+	.resource	  = s3c_spi_lcm_resource,
+	.dev = {
+		.platform_data = &spi_gpio_cfg,
+	},
+};
+
+static struct gta01bl_machinfo backlight_machinfo = {
+	.default_intensity	= 1,
+	.max_intensity		= 1,
+	.limit_mask		= 1,
+};
+
+static struct resource gta01_bl_resources[] = {
+	[0] = {
+		.start	= GTA01_GPIO_BACKLIGHT,
+		.end	= GTA01_GPIO_BACKLIGHT,
+	},
+};
+
+static struct platform_device gta01_bl_dev = {
+	.name 		= "gta01-bl",
+	.num_resources	= ARRAY_SIZE(gta01_bl_resources),
+	.resource	= gta01_bl_resources,
+	.dev		= {
+		.platform_data = &backlight_machinfo,
+	},
+};
+
+static struct resource gta01_led_resources[] = {
+	[0] = {
+		.start	= GTA01_GPIO_VIBRATOR_ON,
+		.end	= GTA01_GPIO_VIBRATOR_ON,
+	},
+};
+
+static struct platform_device gta01_led_dev = {
+	.name		= "neo1973-vibrator",
+	.num_resources	= ARRAY_SIZE(gta01_led_resources),
+	.resource	= gta01_led_resources,
+};
+
+static struct resource gta01_button_resources[] = {
+	[0] = {
+		.start = GTA01_GPIO_AUX_KEY,
+		.end   = GTA01_GPIO_AUX_KEY,
+	},
+	[1] = {
+		.start = GTA01_GPIO_HOLD_KEY,
+		.end   = GTA01_GPIO_HOLD_KEY,
+	},
+	[2] = {
+		.start = GTA01_GPIO_JACK_INSERT,
+		.end   = GTA01_GPIO_JACK_INSERT,
+	},
+};
+
+static struct platform_device gta01_button_dev = {
+	.name		= "neo1973-button",
+	.num_resources	= ARRAY_SIZE(gta01_button_resources),
+	.resource	= gta01_button_resources,
+};
+
+static struct platform_device gta01_pm_gsm_dev = {
+	.name		= "neo1973-pm-gsm",
+};
+
+/* USB */
+static struct s3c2410_hcd_info gta01_usb_info = {
+	.port[0]	= {
+		.flags	= S3C_HCDFLG_USED,
+	},
+	.port[1]	= {
+		.flags	= 0,
+	},
+};
+
+static void __init gta01_map_io(void)
+{
+	s3c24xx_init_io(gta01_iodesc, ARRAY_SIZE(gta01_iodesc));
+	s3c24xx_init_clocks(12*1000*1000);
+	s3c24xx_init_uarts(gta01_uartcfgs, ARRAY_SIZE(gta01_uartcfgs));
+}
+
+extern int gta_gsm_interrupts;
+
+static irqreturn_t gta01_modem_irq(int irq, void *param)
+{
+	printk(KERN_DEBUG "GSM wakeup interrupt (IRQ %d)\n", irq);
+	gta_gsm_interrupts++;
+	nspy_add(NSPY_TYPE_SPECIAL, '*', jiffies);
+	return IRQ_HANDLED;
+}
+
+static void __init gta02f_machine_init(void)
+{
+	int rc;
+
+	if (system_rev == GTA01v4_SYSTEM_REV ||
+	    system_rev == GTA01Bv2_SYSTEM_REV ||
+	    system_rev == GTA01Bv3_SYSTEM_REV ||
+	    system_rev == GTA01Bv4_SYSTEM_REV) {
+		gta01_udc_cfg.udc_command = gta01_udc_command;
+		gta01_mmc_cfg.ocr_avail = MMC_VDD_32_33;
+	}
+
+	s3c_device_usb.dev.platform_data = &gta01_usb_info;
+	s3c_device_nand.dev.platform_data = &gta01_nand_info;
+	s3c_device_sdi.dev.platform_data = &gta01_mmc_cfg;
+
+	s3c24xx_fb_set_platdata(&gta01_lcd_cfg);
+
+	gta01_mmc_cfg.gpio_detect = 0;
+
+	INIT_WORK(&gta01_udc_vbus_drawer.work, __gta01_udc_vbus_draw);
+	s3c24xx_udc_set_platdata(&gta01_udc_cfg);
+	set_s3c2410ts_info(&gta01_ts_cfg);
+
+#if 0
+	/* TODO: choose GPIO outputs that exist on S3C2410 */
+	/* FIXME: hardcoded WLAN module power-up */
+	s3c2410_gpio_cfgpin(GTA02_CHIP_PWD, S3C2410_GPIO_OUTPUT);
+
+	/* Power is down */
+	s3c2410_gpio_setpin(GTA02_CHIP_PWD, 1);
+	mdelay(100);
+
+	s3c2410_gpio_cfgpin(GTA02_GPIO_nWLAN_RESET, S3C2410_GPIO_OUTPUT);
+	/* Chip is in reset state */
+	s3c2410_gpio_setpin(GTA02_GPIO_nWLAN_RESET, 0);
+	mdelay(100);
+	/* Power is up */
+	s3c2410_gpio_setpin(GTA02_CHIP_PWD, 0);
+	mdelay(100);
+	/* Chip is out of reset */
+	s3c2410_gpio_setpin(GTA02_GPIO_nWLAN_RESET, 1);
+#endif
+
+	/* Set LCD_RESET / XRES to high */
+	s3c2410_gpio_cfgpin(S3C2410_GPC6, S3C2410_GPIO_OUTPUT);
+	s3c2410_gpio_setpin(S3C2410_GPC6, 1);
+
+	/* SPI chip select is gpio output */
+	s3c2410_gpio_cfgpin(S3C2410_GPG3, S3C2410_GPIO_OUTPUT);
+	s3c2410_gpio_setpin(S3C2410_GPG3, 1);
+	platform_device_register(&s3c_device_spi_lcm);
+
+	platform_device_register(&gta01_bl_dev);
+	platform_device_register(&gta01_button_dev);
+	platform_device_register(&gta01_pm_gsm_dev);
+
+	switch (system_rev) {
+	case GTA01v3_SYSTEM_REV:
+	case GTA01v4_SYSTEM_REV:
+		/* just use the default (GTA01_IRQ_PCF50606) */
+		break;
+	case GTA01Bv2_SYSTEM_REV:
+	case GTA01Bv3_SYSTEM_REV:
+		/* just use the default (GTA01_IRQ_PCF50606) */
+		gta01_led_resources[0].start =
+			gta01_led_resources[0].end = GTA01Bv2_GPIO_VIBRATOR_ON;
+		break;
+	case GTA01Bv4_SYSTEM_REV:
+		gta01_pmu_resources[0].start =
+			gta01_pmu_resources[0].end = GTA01Bv4_IRQ_PCF50606;
+		gta01_led_resources[0].start =
+			gta01_led_resources[0].end = GTA01Bv4_GPIO_VIBRATOR_ON;
+		break;
+	}
+	mangle_pmu_pdata_by_system_rev();
+	platform_device_register(&gta01_pmu_dev);
+	platform_device_register(&gta01_led_dev);
+
+	platform_device_register(&gta02_sdio_dev);
+
+	platform_add_devices(gta01_devices, ARRAY_SIZE(gta01_devices));
+
+	s3c2410_pm_init();
+
+	set_irq_type(GTA01_IRQ_MODEM, IRQT_RISING);
+	rc = request_irq(GTA01_IRQ_MODEM, gta01_modem_irq, IRQF_DISABLED,
+			 "modem", NULL);
+	enable_irq_wake(GTA01_IRQ_MODEM);
+	printk(KERN_DEBUG  "Enabled GSM wakeup IRQ %d (rc=%d)\n",
+	       GTA01_IRQ_MODEM, rc);
+}
+
+MACHINE_START(NEO1973_GTA02F, "GTA02")
+	.phys_io	= S3C2410_PA_UART,
+	.io_pg_offst	= (((u32)S3C24XX_VA_UART) >> 18) & 0xfffc,
+	.boot_params	= S3C2410_SDRAM_PA + 0x100,
+	.map_io		= gta01_map_io,
+	.init_irq	= s3c24xx_init_irq,
+	.init_machine	= gta02f_machine_init,
+	.timer		= &s3c24xx_timer,
+MACHINE_END
diff --git a/arch/arm/mach-s3c2440/mach-gta02.c b/arch/arm/mach-s3c2440/mach-gta02.c
index 9744ac2..0a12297 100644
--- a/arch/arm/mach-s3c2440/mach-gta02.c
+++ b/arch/arm/mach-s3c2440/mach-gta02.c
@@ -376,12 +376,12 @@ int gta02_get_pcb_revision(void)
 	return u;
 }
 
-struct platform_device gta02_version_device = {
+static struct platform_device gta02_version_device = {
 	.name 		= "neo1973-version",
 	.num_resources	= 0,
 };
 
-struct platform_device gta02_resume_reason_device = {
+static struct platform_device gta02_resume_reason_device = {
 	.name 		= "neo1973-resume",
 	.num_resources	= 0,
 };
@@ -631,7 +631,7 @@ static struct resource gta02_pmu_resources[] = {
 	},
 };
 
-struct platform_device gta02_pmu_dev = {
+static struct platform_device gta02_pmu_dev = {
 	.name 		= "pcf50633",
 	.num_resources	= ARRAY_SIZE(gta02_pmu_resources),
 	.resource	= gta02_pmu_resources,
@@ -650,7 +650,7 @@ static struct resource sc32440_fiq_resources[] = {
 	},
 };
 
-struct platform_device sc32440_fiq_device = {
+static struct platform_device sc32440_fiq_device = {
 	.name 		= "sc32440_fiq",
 	.num_resources	= 1,
 	.resource	= sc32440_fiq_resources,
@@ -666,7 +666,7 @@ static struct resource gta02_hdq_resources[] = {
 	},
 };
 
-struct platform_device gta02_hdq_device = {
+static struct platform_device gta02_hdq_device = {
 	.name 		= "gta02-hdq",
 	.num_resources	= 1,
 	.resource	= gta02_hdq_resources,
@@ -675,7 +675,7 @@ struct platform_device gta02_hdq_device = {
 
 /* BQ27000 Battery */
 
-struct bq27000_platform_data bq27000_pdata = {
+static struct bq27000_platform_data bq27000_pdata = {
 	.name = "bat",
 	.rsense_mohms = 20,
 	.hdq_read = gta02hdq_read,
@@ -683,7 +683,7 @@ struct bq27000_platform_data bq27000_pdata = {
 	.hdq_initialized = gta02hdq_initialized,
 };
 
-struct platform_device bq27000_battery_device = {
+static struct platform_device bq27000_battery_device = {
 	.name 		= "bq27000-battery",
 	.dev = {
 		.platform_data = &bq27000_pdata,
@@ -747,7 +747,7 @@ static struct platform_device gta02_sdio_dev = {
         .num_resources  = ARRAY_SIZE(gta02_sdio_resources),
 };
 
-struct platform_device s3c24xx_pwm_device = {
+static struct platform_device s3c24xx_pwm_device = {
 	.name 		= "s3c24xx_pwm",
 	.num_resources	= 0,
 };
@@ -852,19 +852,19 @@ static struct s3c2410_ts_mach_info gta02_ts_cfg = {
 
 /* SPI: LCM control interface attached to Glamo3362 */
 
-void gta02_jbt6k74_reset(int devidx, int level)
+static void gta02_jbt6k74_reset(int devidx, int level)
 {
 	glamo_lcm_reset(level);
 }
 
 /* finally bring up deferred backlight resume now LCM is resumed itself */
 
-void gta02_jbt6k74_resuming(int devidx)
+static void gta02_jbt6k74_resuming(int devidx)
 {
 	pcf50633_backlight_resume(pcf50633_global);
 }
 
-const struct jbt6k74_platform_data jbt6k74_pdata = {
+static const struct jbt6k74_platform_data jbt6k74_pdata = {
 	.reset		= gta02_jbt6k74_reset,
 	.resuming	= gta02_jbt6k74_resuming,
 };
@@ -1156,7 +1156,7 @@ static struct resource gta02_led_resources[] = {
 	},
 };
 
-struct platform_device gta02_led_dev = {
+static struct platform_device gta02_led_dev = {
 	.name		= "gta02-led",
 	.num_resources	= ARRAY_SIZE(gta02_led_resources),
 	.resource	= gta02_led_resources,
diff --git a/arch/arm/tools/mach-types b/arch/arm/tools/mach-types
index 0a9a5e7..f2d90b0 100644
--- a/arch/arm/tools/mach-types
+++ b/arch/arm/tools/mach-types
@@ -1367,3 +1367,4 @@ db88f5281		MACH_DB88F5281		DB88F5281		1358
 csb726			MACH_CSB726		CSB726			1359
 tik27			MACH_TIK27		TIK27			1360
 mx_uc7420		MACH_MX_UC7420		MX_UC7420		1361
+neo1973_gta02f		MACH_NEO1973_GTA02F	NEO1973_GTA02F		1555
diff --git a/defconfig-2.6.22.5 b/defconfig-2.6.22.5
index e9e40fe..71196a1 100644
--- a/defconfig-2.6.22.5
+++ b/defconfig-2.6.22.5
@@ -180,6 +180,7 @@ CONFIG_S3C2410_PWM=y
 # CONFIG_MACH_VR1000 is not set
 CONFIG_MACH_QT2410=y
 CONFIG_MACH_NEO1973_GTA01=y
+CONFIG_MACH_NEO1973_GTA02F=y
 
 #
 # S3C2412 Machines
diff --git a/defconfig-2.6.24 b/defconfig-2.6.24
index 0bcf50a..cf13b66 100644
--- a/defconfig-2.6.24
+++ b/defconfig-2.6.24
@@ -187,6 +187,7 @@ CONFIG_S3C2410_PWM=y
 # CONFIG_MACH_VR1000 is not set
 CONFIG_MACH_QT2410=y
 CONFIG_MACH_NEO1973_GTA01=y
+CONFIG_MACH_NEO1973_GTA02F=y
 
 #
 # S3C2412 Machines
diff --git a/defconfig-gta02 b/defconfig-gta02
index 68cf1f4..17cb17b 100644
--- a/defconfig-gta02
+++ b/defconfig-gta02
@@ -184,6 +184,7 @@ CONFIG_S3C2410_PWM=y
 # CONFIG_MACH_VR1000 is not set
 CONFIG_MACH_QT2410=y
 CONFIG_MACH_NEO1973_GTA01=y
+CONFIG_MACH_NEO1973_GTA02F=y
 
 #
 # S3C2412 Machines
diff --git a/drivers/sdio/hcd/s3c24xx/s3c24xx_hcd.c b/drivers/sdio/hcd/s3c24xx/s3c24xx_hcd.c
index 1292020..33577ac 100644
--- a/drivers/sdio/hcd/s3c24xx/s3c24xx_hcd.c
+++ b/drivers/sdio/hcd/s3c24xx/s3c24xx_hcd.c
@@ -100,7 +100,7 @@ static void s3c24xx_dump_regs(struct s3c24xx_hcd_context * context)
 	datcnt 	= readl(context->base + S3C2410_SDIDCNT);
 	datsta 	= readl(context->base + S3C2410_SDIDSTA);
 	fsta 	= readl(context->base + S3C2410_SDIFSTA);
-	imask   = readl(context->base + S3C2440_SDIIMSK);
+	imask   = readl(context->base + S3C2410_SDIIMSK);
 
 	printk("SDICON:    0x%08x\n", con);
 	printk("SDIPRE:    0x%08x\n", pre);
@@ -124,15 +124,15 @@ static inline void s3c24xx_hcd_clear_imask(struct s3c24xx_hcd_context * context)
 {
 	if (context->int_sdio) {
 		writel(S3C2410_SDIIMSK_SDIOIRQ | S3C2410_SDIIMSK_READWAIT,
-		       context->base + S3C2440_SDIIMSK);
+		       context->base + S3C2410_SDIIMSK);
 	}  else {
-		writel(0, context->base + S3C2440_SDIIMSK);
+		writel(0, context->base + S3C2410_SDIIMSK);
 	}
 }
 
 static inline void s3c24xx_hcd_set_imask(struct s3c24xx_hcd_context * context)
 {
-	writel(context->int_mask, context->base + S3C2440_SDIIMSK);
+	writel(context->int_mask, context->base + S3C2410_SDIIMSK);
 }
 
 
@@ -453,11 +453,11 @@ static void s3c24xx_hcd_pio_complete(struct s3c24xx_hcd_context * context)
 			req->DataRemaining -= fifo_count;
 			while (fifo_count > 0) {
 				if (context->data_size == 4)
-					*(ptr) = readl(context->base + S3C2440_SDIDATA);
+					*(ptr) = readl(context->base + S3C2410_SDIDATA);
 				else if (context->data_size == 2)
-					*(ptr) = readw(context->base + S3C2440_SDIDATA);
+					*(ptr) = readw(context->base + S3C2410_SDIDATA);
 				else
-					*(ptr) = readb(context->base + S3C2440_SDIDATA);
+					*(ptr) = readb(context->base + S3C2410_SDIDATA);
 
 				ptr += context->data_size;
 				fifo_count -= context->data_size;
@@ -521,11 +521,11 @@ static void s3c24xx_hcd_pio_complete(struct s3c24xx_hcd_context * context)
 
 				while (fifo_count > 0) {
 					if (context->data_size == 4)
-						writel(*(ptr), context->base + S3C2440_SDIDATA);
+						writel(*(ptr), context->base + S3C2410_SDIDATA);
 					else if (context->data_size == 2)
-						writew(*(ptr), context->base + S3C2440_SDIDATA);
+						writew(*(ptr), context->base + S3C2410_SDIDATA);
 					else
-						writeb(*(ptr), context->base + S3C2440_SDIDATA);
+						writeb(*(ptr), context->base + S3C2410_SDIDATA);
 
 					ptr += context->data_size;
 					fifo_count -= context->data_size;
@@ -709,7 +709,7 @@ static int s3c24xx_hcd_prepare_dma(struct s3c24xx_hcd_context * context)
 	}
 
 	s3c2410_dma_devconfig(context->dma_channel, source, hwcfg,
-			      (unsigned long)context->mem->start + S3C2440_SDIDATA);
+			      (unsigned long)context->mem->start + S3C2410_SDIDATA);
 
 	s3c2410_dma_config(context->dma_channel, context->data_size,
 			   S3C2410_DCON_CH0_SDI);
@@ -759,9 +759,9 @@ static irqreturn_t s3c24xx_hcd_irq(int irq, void *dev_id)
 
 			context->int_sdio = 0;
 
-			imask = readl(context->base + S3C2440_SDIIMSK);
+			imask = readl(context->base + S3C2410_SDIIMSK);
 			imask &= ~S3C2410_SDIIMSK_SDIOIRQ;
-			writel(imask, context->base + S3C2440_SDIIMSK);
+			writel(imask, context->base + S3C2410_SDIIMSK);
 			schedule_work(&context->irq_work);
 		}
 	}
@@ -895,20 +895,20 @@ SDIO_STATUS s3c24xx_hcd_config(PSDHCD hcd, PSDCONFIG config)
 
 		if (int_data->SlotIRQEnable &
 		    (IRQ_DETECT_1_BIT | IRQ_DETECT_4_BIT | IRQ_DETECT_MULTI_BLK) ) {
-			imsk = readl(context->base + S3C2440_SDIIMSK);
+			imsk = readl(context->base + S3C2410_SDIIMSK);
 
 			if (int_data->SlotIRQEnable) {
 				printk("SDIO_INT_CTRL enable IRQ\n");
 				DBG_PRINT(SDDBG_TRACE, ("SDIO_INT_CTRL enable IRQ\n"));
 				context->int_sdio = 1;
 				imsk |= S3C2410_SDIIMSK_SDIOIRQ;
-				writel(imsk, context->base + S3C2440_SDIIMSK);
+				writel(imsk, context->base + S3C2410_SDIIMSK);
 			} else {
 				printk("SDIO_INT_CTRL disable IRQ\n");
 				DBG_PRINT(SDDBG_TRACE, ("SDIO_INT_CTRL disable IRQ\n"));
 				context->int_sdio = 0;
 				imsk &= ~S3C2410_SDIIMSK_SDIOIRQ;
-				writel(imsk, context->base + S3C2440_SDIIMSK);
+				writel(imsk, context->base + S3C2410_SDIIMSK);
 			}
 		}
 		status = SDIO_STATUS_SUCCESS;
@@ -917,9 +917,9 @@ SDIO_STATUS s3c24xx_hcd_config(PSDHCD hcd, PSDCONFIG config)
 		DBG_PRINT(SDDBG_TRACE, ("config SDIO_REARM_INT\n"));
 
 		context->int_sdio = 1;
-		imsk = readl(context->base + S3C2440_SDIIMSK);
+		imsk = readl(context->base + S3C2410_SDIIMSK);
 		imsk |= S3C2410_SDIIMSK_SDIOIRQ;
-		writel(imsk, context->base + S3C2440_SDIIMSK);
+		writel(imsk, context->base + S3C2410_SDIIMSK);
 
 		status = SDIO_STATUS_SUCCESS;
 		break;
@@ -1110,7 +1110,7 @@ SDIO_STATUS s3c24xx_hcd_request(PSDHCD hcd)
 	if (context->int_sdio)
 		imask |= S3C2410_SDIIMSK_SDIOIRQ;
 	context->int_mask = imask;
-	writel(imask, context->base + S3C2440_SDIIMSK);
+	writel(imask, context->base + S3C2410_SDIIMSK);
 	writel(req->Argument, context->base + S3C2410_SDICMDARG);
 	writel(cmdcon, context->base + S3C2410_SDICMDCON);
 
@@ -1388,7 +1388,7 @@ static int s3c24xx_hcd_suspend(struct platform_device * pdev, pm_message_t state
 	context->suspend_regs.datcnt 	= readl(context->base + S3C2410_SDIDCNT);
 	context->suspend_regs.datsta 	= readl(context->base + S3C2410_SDIDSTA);
 	context->suspend_regs.fsta 	= readl(context->base + S3C2410_SDIFSTA);
-	context->suspend_regs.imask   = readl(context->base + S3C2440_SDIIMSK);
+	context->suspend_regs.imask   = readl(context->base + S3C2410_SDIIMSK);
 
 	spin_unlock_irqrestore(&context->lock, flags);
 	return 0;
@@ -1416,7 +1416,7 @@ static int s3c24xx_hcd_resume(struct platform_device * pdev)
 	writel(context->suspend_regs.datcnt, context->base + S3C2410_SDIDCNT);
 	writel(context->suspend_regs.datsta, context->base + S3C2410_SDIDSTA);
 	writel(context->suspend_regs.fsta, context->base + S3C2410_SDIFSTA);
-	writel(context->suspend_regs.imask, context->base + S3C2440_SDIIMSK);
+	writel(context->suspend_regs.imask, context->base + S3C2410_SDIIMSK);
 
 	spin_unlock_irqrestore(&context->lock, flags);
 	return 0;
@@ -1462,7 +1462,7 @@ static int s3c24xx_hcd_debugfs_show(struct seq_file *s, void *data)
 	datcnt 	= readl(context->base + S3C2410_SDIDCNT);
 	datsta 	= readl(context->base + S3C2410_SDIDSTA);
 	fsta 	= readl(context->base + S3C2410_SDIFSTA);
-	imask   = readl(context->base + S3C2440_SDIIMSK);
+	imask   = readl(context->base + S3C2410_SDIIMSK);
 
 	seq_printf(s, "SDICON:    0x%08x\n", con);
 	seq_printf(s, "SDIPRE:    0x%08x\n", pre);
diff --git a/include/asm-arm/plat-s3c24xx/pm.h b/include/asm-arm/plat-s3c24xx/pm.h
index cc62366..20e03b3 100644
--- a/include/asm-arm/plat-s3c24xx/pm.h
+++ b/include/asm-arm/plat-s3c24xx/pm.h
@@ -16,6 +16,8 @@
 
 #ifdef CONFIG_PM
 
+#include <linux/sysdev.h>
+
 extern __init int s3c2410_pm_init(void);
 
 #else
